### PR TITLE
Laboratory work #1, Karabasova Julia - 19FPL2

### PR DESCRIPTION
**Pull request by juliakarabasova**

### CODE REVIEW

File: lab_1/stop_words.txt [hunk 1]
```diff
@@ -71,6 +71,7 @@ better
 between
 beyond
 both
+br
```
Comment (dmitry-uraev) [inline lab_1/stop_words.txt]: I do not understand this addition. Can you explain?

File: lab_1/stop_words.txt [hunk 2]
```diff
@@ -541,4 +542,4 @@ your
 yours
 yourself
 yourselves
-zero
```
Comment (dmitry-uraev) [inline lab_1/stop_words.txt]: Why replace with the same?

**Comment by dmitry-uraev** [comment]
1. Consider adding start.py realisation.
2. Fix lint.

File: lab_1/stop_words.txt (hunk 1 above)
Comment (juliakarabasova) [reply lab_1/stop_words.txt]: That was for the case if one of tests (marked with @unittestskip or so) starts working. I guess I should delete it

File: lab_1/stop_words.txt (hunk 2 above)
Comment (juliakarabasova) [reply lab_1/stop_words.txt]: I just deleted the last empty line, but if its forbidden I can realise it in the code.

File: lab_1/stop_words.txt (hunk 1 above)
Comment (dmitry-uraev) [reply lab_1/stop_words.txt]: That's true. We removed that test.

File: lab_1/stop_words.txt (hunk 2 above)
Comment (dmitry-uraev) [reply lab_1/stop_words.txt]: Better not to touch given files. Consider them as untouchables.

File: lab_1/main.py [hunk 3]
```diff
@@ -12,7 +15,10 @@ def tokenize(text: str) -> list:
     e.g. text = 'The weather is sunny, the man is happy.'
     --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     """
-    pass
+    if not isinstance(text, str):
+        return []
+    text_output = re.sub('[^a-z \n]', '', text.lower()).split()
```
Comment (dmitry-uraev) [inline lab_1/main.py:20]: be prepared to implement this function without re if asked on October 3th

File: lab_1/main.py [hunk 4]
```diff
@@ -25,7 +31,10 @@ def remove_stop_words(tokens: list, stop_words: list) -> list:
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    if not isinstance(tokens, list):
+        return []
+    list_words = [word for word in tokens if word not in stop_words]
+    return list_words
```
Comment (dmitry-uraev) [inline lab_1/main.py:37]: nice!

File: lab_1/main.py [hunk 5]
```diff
@@ -36,7 +45,13 @@ def calculate_frequencies(tokens: list) -> dict:
...
     --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
     """
-    pass
+    if not isinstance(tokens, list):
+        return {}
+    if len(tokens) > 0 and not isinstance(tokens[0], str):
+        return {}
+    set_words = set(tokens.copy())
+    dict_freq = {word: tokens.count(word) for word in set_words}
+    return dict_freq
```
Comment (dmitry-uraev) [inline lab_1/main.py:54]: +

File: lab_1/main.py [hunk 6]
```diff
@@ -49,7 +64,10 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     top_n = 1
     --> ['happy']
     """
-    pass
+    if not isinstance(freq_dict, dict) or not isinstance(top_n, int):
+        return []
+    list_output = sorted(freq_dict, key=freq_dict.get, reverse=True)
+    return list_output[:top_n]
```
Comment (dmitry-uraev) [inline lab_1/main.py:70]: +

File: lab_1/main.py [hunk 7]
```diff
@@ -69,7 +87,33 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
...
+    if right_context_size == 0 and left_context_size == 0:
+        return []
+    if (indexes[-1] + right_context_size) > len(tokens):
+        right_context_size = len(tokens)
+
+    if (indexes[0] - left_context_size) < 0:
+        list_output = [tokens[0:ind + 1 + right_context_size] for ind in indexes]
+    else:
+        list_output = [tokens[ind - left_context_size:ind + 1 + right_context_size] for ind in indexes]
+    return list_output
```
Comment (dmitry-uraev) [inline lab_1/main.py:116]: suggest a shorter implementation. Be ready to show on October 3th

File: lab_1/main.py [hunk 8]
```diff
@@ -125,4 +180,18 @@ def sort_concordance(tokens: list, word: str, left_context_size: int, right_cont
...
+        right_context_size = 0
+    concordance = get_concordance(tokens, word, left_context_size, right_context_size)
+    if len(concordance) == 0:
+        return []
+    if left_sort:
+        dict_raw = {context[0]: context for context in concordance}
+    else:
+        dict_raw = {context[context.index(word)+1]: context for context in concordance}
+    list_output = [dict_raw[key] for key in sorted(dict_raw)]
+    return list_output
```
Comment (dmitry-uraev) [inline lab_1/main.py:200]: Nice!

File: lab_1/start.py [hunk 9]
```diff
@@ -2,17 +2,42 @@
...
+    if len(sorted_concordance_right) > 0:
+        print('Concordance sorted by the first right word (first 5 cases):')
+        for concordance in sorted_concordance_right[:5]:
+            print('\t', concordance)
 
-    RESULT = None
+    RESULT = sorted_concordance_left
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
-    assert RESULT == [(), ()], 'Concordance not working'
+    assert RESULT != [], 'Concordance not working'
```
Comment (dmitry-uraev) [inline lab_1/start.py]: just keep `assert RESULT` here

**Comment by dmitry-uraev** [comment]
Overall comment:
Great!
Nice styling. I like it.
Just fix some small additions.

**Pull request merged**